primitives/src/merkle_tree/namespaced_merkle_tree/mod.rs

```diff
@@ -221,22 +217,24 @@ where
         &mut self,
         elems: impl IntoIterator<Item = impl core::borrow::Borrow<Self::Element>>,
     ) -> Result<(), PrimitivesError> {
-        // TODO(#314): update namespace metadata
-        self.inner.extend(elems)
+        let leaves =
+            NMT::<E, H, Arity, N, T>::update_namespace_metadata(&mut self.namespace_ranges, elems)?;
+        self.inner.extend(leaves)
     }
 
     fn push(
         &mut self,
         elem: impl core::borrow::Borrow<Self::Element>,
     ) -> Result<(), PrimitivesError> {
-        // TODO(#314): update namespace metadata
-        self.inner.push(elem)
+        let cloned = elem.borrow().clone();
```

Comment:
In practice why do we need to update the tree? Is it not enough to build the tree from the leaves only once? Or do we do this in order to keep the consistency of the API?

Reply:
The current Hotshot Block trait requires us to push one tx at a time

Reply:
@jbearer it would be ideal if we could construct the HotShot block in one go @jbearer does something about consensus preclude us from doing this?

Reply:
It comes from validating consensus, where the builder would push one transaction at a time, checking if it made the block inconsistent, as a way to filter out invalid transactions from the proposal. I think for sequencing consensus the builder could theoretically collect all transactions into a `Vec` as they are submitted, and then build the block all at once. You'll have to talk to the consensus team to see if an API change is possible